### ISSUE TITLE
Control Plane Lock should record which machine has been elected

### ISSUE THREAD

**Issue by liztio** [kubernetes-sigs/cluster-api-provider-aws#981]
/kind feature

**Describe the solution you'd like**
Right now, the Control Plane Lock does not document what machine has been chosen to be the first control plane (the one that gets `kubeadm init` instead of `kubeadm join`).

If we recorded this, we could have better log messages, record events, and potentially even restart a failed machine creation.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-aws version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

**Comment by liztio** [comment]
Proposed flow:
1. Write out a configmap that references the machine that was elected
2. Requeue immediately
3. Retrieve Machine listed in ConfigMap
4. Call GetOrCreate()
5. Update [Instance](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/release-0.3/pkg/apis/awsprovider/v1alpha1/types.go#L411)
6. Requeue if not ready

**Comment by liztio** [comment]
GetOrCreate() will reacquire any orphaned EC2 instances
The arl will be populated immediately for easier debugging

I'm worried about duplicating instances during 4) though

**Comment by ncdc** [comment]
> 2. Requeue immediately

Are you suggesting this instead of allowing the winner to immediately proceed so that we only have a single code path & flow to test?

**Comment by liztio** [comment]
It's so we can be sure that the etcd commit was actually successful

**Comment by ncdc** [comment]
commit of what resource?

**Comment by liztio** [comment]
the lock

**Comment by ncdc** [comment]
`ConfigMaps(foo).Create(&cm)` will only return a `nil` error if it succeeded. We don't need to requeue to ensure the ConfigMap was written to etcd.

**Comment by liztio** [comment]
Fair enough, just being defensive

**Comment by ncdc** [comment]
It _could_ be easier to test if we do what you're suggesting. Otherwise, we do have 2 flows:

1. Create lock, proceed because it was successful
1. Handle case where lock was created but something went wrong and we didn't get a happy VM (i.e. proceed b/c Machine name/uid matches what's in the lock)

If we do what you're suggesting, it's:

1. Iteration 1: create lock
1. Iteration 2: query lock & proceed if lock details match current Machine

right?

**Comment by liztio** [comment]
Yeah, exactly. I don't have a sense of how long our backlog queue usually is, but if it's long on a busy server splitting up could add increased latency

**Comment by liztio** [comment]
We could write an idempotency token to the configmap to keep step 4) from being deleted

**Comment by ncdc** [comment]
We currently reconcile 1 machine at a time, and given this is strictly about the initialization of the control plane, anything else that were to get reconciled after we exit iteration 1 would either quickly be requeued (wrong machine) or quickly proceed (correct machine).

**Comment by ncdc** [comment]
It's ok to call GetOrCreate multiple times, as long as it only ever creates once. Do you think there is anything wrong with GetOrCreate that would cause the Get portion of it to fail after we have an ec2 instance?

**Comment by detiber** [comment]
@ncdc There is the possibility of a small window related to ec2 eventual consistency where we create a new instance, the controller dies or for some other reason, the wait for instance ready fails to complete, and the requeue happens fast enough that the subsequent describe call fails to return the previously created instance.

**Comment by liztio** [comment]
I'm worried about the eventual consistency issues @detiber described, where we would've already called create but the machine isn't quite available in the instance

**Comment by liztio** [comment]
![image](https://user-images.githubusercontent.com/33235082/62785274-54a91400-ba8e-11e9-98fb-25b386278d5c.png)

**Comment by liztio** [comment]
![image](https://user-images.githubusercontent.com/33235082/62790228-a3f44200-ba98-11e9-86a6-9fab6cc02b2e.png)

**Comment by liztio** [comment]
`aws-provide…┊ I0812 13:03:43.068069       1 control_plane_init_locker.go:81] [machine-actuator] "level"=0 "msg"="waiting on on another machine to initialize" "cluster-name"="ponyville" "configmap-name"="009cd091-bd01-11e9-a06a-0242ac110002-controlplane" "machine-name"="ponyville-controlplane-2" "namespace"="equestria" "init-machine"="ponyville-controlplane-0"`
